source/agora/test/EnrollmentManager.d

```diff
@@ -95,3 +95,135 @@ unittest
+///     information of the new block is adding to the validator set.
+/// Expectation: The node restores its own enrollment data when the block
+///     containing the enrollment information arrives.
+unittest
+{
+    import agora.consensus.validation.PreImage;
+    import core.thread;
+    import std.conv;
+
+    auto network = makeTestNetwork(TestConf.init);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+
+    // Boilerplate
+    auto nodes = network.clients;
+    auto gen_key_pair = getGenesisKeyPair();
+    auto node_1 = nodes[0];
+    auto pubkey_1 = node_1.getPublicKey();
+
+    // Create a new block with frozen transactions so we can Enroll
+    Transaction[] txs;
+    {
+        foreach (idx; 0 .. Block.TxsInBlock)
+        {
+            auto input = Input(hashFull(GenesisTransaction), idx.to!uint);
+            Transaction tx =
+                {
+                    TxType.Freeze,
+                    [input],
+                    [Output(Amount.MinFreezeAmount, pubkey_1),
+                        Output(Amount(1000), gen_key_pair.address)]
+                };
+            auto signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+            tx.inputs[0].signature = signature;
+            txs ~= tx;
+        }
+        // Send all transactions to node 1 and rely on gossip
+        txs.each!(tx => node_1.putTransaction(tx));
+        // Make sure the block is created
+        containSameBlocks(nodes, 1).retryFor(5.seconds);
+    }
+
+    // The nodes[0] register as validator
+    auto enroll = node_1.createEnrollmentData();
+    node_1.enrollValidator(enroll);
+    nodes.each!(node =>
+        retryFor(node.getEnrollment(enroll.utxo_key) == enroll, 5.seconds));
+
+    // The `enroll` hasn't been validator yet so there must not be a pre-image
+    assert(node_1.getPreimage(enroll.utxo_key) == PreImageInfo.init);
+
+    // Check if the `node_1` restores block from the chain after restarting
+    node_1.restart();
+    RemoteAPI!TestAPI[] new_nodes;
+    new_nodes ~= node_1;
+    foreach (node; nodes)
+    {
+        writeln("getHeight of node:", node.getBlockHeight());
+        new_nodes ~= node;
+    }
+    containSameBlocks(new_nodes, 1).retryFor(5.seconds);
+
+    // Now make a second block, and check what's inside
+    Transaction[] txs2;
+    {
+        foreach (idx; 0 .. Block.TxsInBlock)
+        {
+            auto input = Input(hashFull(txs[idx]), 1);
+
+            Transaction tx =
+            {
+                TxType.Payment,
+                [input],
+                [Output(Amount(100), gen_key_pair.address)]
+            };
+
+            auto signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+            tx.inputs[0].signature = signature;
+            txs2 ~= tx;
+        }
+        txs2.each!(tx => node_1.putTransaction(tx));
+        containSameBlocks(nodes, 2).retryFor(20.seconds);
```

Comment:
This doesn't quite test what we discussed. It's a good test, for sure, but what we discussed was to create a test where the validator gets its Enrollment during *catchup*. Here it gets it through regular blockchain update.